### PR TITLE
core: consider non-SERVICE_EXEC_START commands for EXIT_CLEAN_COMMAND

### LINKED ISSUE #11431: TimeoutStopSec= documentation contradicts actual behaviour
**systemd version the issue has been seen with**
> 240

**Used distribution**
 > Debian

**Expected behaviour you didn't see**
 > TimeoutStopSec= expiring on first ExecStop= skips subsequent ExecStop= statements

**Unexpected behaviour you saw**
 > The next ExecStop= statement is executed as the control process, while the service is already transitioned to stop-sigterm stage.

**Steps to reproduce the problem**
```ini
% cat .bar.service
[Unit]
Description=bar
[Service]
ExecStart=/bin/sleep infinity
TimeoutStopSec=2s
ExecStop=/bin/sh -c "echo number1; exec /bin/sleep 5"
ExecStop=/bin/sh -c "echo number2; exec /bin/sleep 5"
ExecStop=/bin/sh -c "echo number3; exec /bin/sleep 5"
ExecStop=/bin/sh -c "echo number4; exec /bin/sleep 5"
ExecStop=/bin/sh -c "echo number5; exec /bin/sleep 5"
```

```
Jan 15 09:07:52 jupiter systemd[1257]: Stopping bar...
Jan 15 09:07:52 jupiter systemd[1257]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/job/3501 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7181 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Jan 15 09:07:52 jupiter systemd[1257]: Got message type=method_call sender=:1.1915 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=GetUnit cookie=4 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Jan 15 09:07:52 jupiter systemd[1257]: Sent message type=method_return sender=n/a destination=:1.1915 path=n/a interface=n/a member=n/a cookie=7182 reply_cookie=4 signature=o error-name=n/a error-message=n/a
Jan 15 09:07:52 jupiter systemd[1257]: Got message type=method_call sender=:1.1915 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/bar_2eservice interface=org.freedesktop.DBus.Properties member=Get cookie=5 reply_cookie=0 signature=ss error-name=n/a error-message=n/a
Jan 15 09:07:52 jupiter sh[9080]: number1
Jan 15 09:07:52 jupiter systemd[9080]: Skipping PR_SET_MM, as we don't have privileges.
Jan 15 09:07:52 jupiter systemd[1257]: Sent message type=method_return sender=n/a destination=:1.1915 path=n/a interface=n/a member=n/a cookie=7183 reply_cookie=5 signature=v error-name=n/a error-message=n/a
Jan 15 09:07:52 jupiter systemd[9080]: bar.service: Executing: /bin/sh -c 'echo number1; exec /bin/sleep 5'
Jan 15 09:07:54 jupiter systemd[1257]: bar.service: Stopping timed out. Terminating.
Jan 15 09:07:54 jupiter systemd[1257]: bar.service: Changed stop -> stop-sigterm
Jan 15 09:07:54 jupiter systemd[1257]: Received SIGCHLD from PID 9080 (sleep).
Jan 15 09:07:54 jupiter systemd[1257]: Child 9077 (sleep) died (code=killed, status=15/TERM)
Jan 15 09:07:54 jupiter systemd[1257]: bar.service: Child 9077 belongs to bar.service.
Jan 15 09:07:54 jupiter systemd[1257]: bar.service: Main process exited, code=killed, status=15/TERM
Jan 15 09:07:54 jupiter systemd[1257]: Child 9080 (sleep) died (code=killed, status=15/TERM)
Jan 15 09:07:54 jupiter systemd[1257]: bar.service: Child 9080 belongs to bar.service.
Jan 15 09:07:54 jupiter systemd[1257]: bar.service: Control process exited, code=killed, status=15/TERM
Jan 15 09:07:54 jupiter systemd[1257]: bar.service: Running next control command for state stop-sigterm.
Jan 15 09:07:54 jupiter systemd[1257]: bar.service: About to execute: /bin/sh -c 'echo number2; exec /bin/sleep 5'
Jan 15 09:07:54 jupiter systemd[1257]: bar.service: Forked /bin/sh as 9081
Jan 15 09:07:54 jupiter systemd[9081]: Skipping PR_SET_MM, as we don't have privileges.
Jan 15 09:07:54 jupiter systemd[9081]: bar.service: Executing: /bin/sh -c 'echo number2; exec /bin/sleep 5'
Jan 15 09:07:54 jupiter sh[9081]: number2
Jan 15 09:07:56 jupiter systemd[1257]: bar.service: State 'stop-sigterm' timed out. Killing.
Jan 15 09:07:56 jupiter systemd[1257]: bar.service: Killing process 9081 (sleep) with signal SIGKILL.
Jan 15 09:07:56 jupiter systemd[1257]: bar.service: Changed stop-sigterm -> stop-sigkill
Jan 15 09:07:56 jupiter systemd[1257]: Received SIGCHLD from PID 9081 (sleep).
Jan 15 09:07:56 jupiter systemd[1257]: Child 9081 (sleep) died (code=killed, status=9/KILL)
Jan 15 09:07:56 jupiter systemd[1257]: bar.service: Child 9081 belongs to bar.service.
Jan 15 09:07:56 jupiter systemd[1257]: bar.service: Control process exited, code=killed, status=9/KILL
Jan 15 09:07:56 jupiter systemd[1257]: bar.service: Got final SIGCHLD for state stop-sigkill.
Jan 15 09:07:56 jupiter systemd[1257]: bar.service: Failed with result 'timeout'.
Jan 15 09:07:56 jupiter systemd[1257]: bar.service: Changed stop-sigkill -> failed
Jan 15 09:07:56 jupiter systemd[1257]: bar.service: Job 3501 bar.service/stop finished, result=done
Jan 15 09:07:56 jupiter systemd[1257]: Stopped bar.
```

number2 should have never been invoked...

### PR DESCRIPTION
**Pull request by anitazha**
When there are multiple ExecStop= statements, the next command would continue
to run even after TimeoutStopSec= is up and sends SIGTERM. This is because,
unless Type= is oneshot, the exit code/status would evaluate to SERVICE_SUCCESS
in service_sigchld_event()'s call to is_clean_exit(). This success indicates
following commands would continue running until the end of the list
is reached, or another timeout is hit and SIGKILL is sent.

Since long running processes should not be invoked in non-SERVICE_EXEC_START
commands, consider them for EXIT_CLEAN_COMMAND instead of EXIT_CLEAN_DAEMON.
Passing EXIT_CLEAN_COMMAND to is_clean_exit() evaluates the SIGTERM exit
code/status to failure and will stop execution after the first timeout is hit.

Fixes #11431

### CODE REVIEW

File: src/core/service.c [hunk 1]
```diff
@@ -3214,11 +3214,19 @@ static void service_sigchld_event(Unit *u, pid_t pid, int code, int status) {
...
         ServiceResult f;
+        ExitClean clean_mode;
 
         assert(s);
         assert(pid >= 0);
 
-        if (is_clean_exit(code, status, s->type == SERVICE_ONESHOT ? EXIT_CLEAN_COMMAND : EXIT_CLEAN_DAEMON, &s->success_status))
+        /* Oneshot services and non-SERVICE_EXEC_START commands should not be
+         * considered daemons as they are typically not long running. */
+        if ((s->type == SERVICE_ONESHOT) || (s->control_command_id != SERVICE_EXEC_START))
```
Comment (poettering) [inline src/core/service.c]: nitpick: the inner two pairs of `()` are unnecessary. In C == and != bind stronger than ||...

Comment (poettering) [reply src/core/service.c]: Hmm, also, the PID of the child which died needs to be checked before assuming that this is a control command... i.e.

```c
…
if (s->type == SERVICE_ONESHOT || 
        (pid == s->control_pid && s->control_command_id != SERVICE_EXEC_START))
…
```

**Comment by anitazha** [comment]
Updated with fixes. Please take a look.

**Comment by poettering** [comment]
Looks great, thanks!

**Pull request merged**